orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/cloudformation/WaitForCloudFormationCompletionTask.java

```diff
@@ -122,6 +115,22 @@ public long getTimeout() {
     return timeout;
   }
 
+  private String getFailureReason(Map stack, StageExecution stage) {
+    List unrecoverableStatuses =
+        Arrays.asList(
+            CloudFormationStates.ROLLBACK_COMPLETE.name(),
+            CloudFormationStates.ROLLBACK_FAILED.name(),
+            CloudFormationStates.DELETE_FAILED.name(),
+            CloudFormationStates.UPDATE_ROLLBACK_FAILED.name());
+    if (unrecoverableStatuses.contains(stack.get("stackStatus"))) {
+      return "Irrecoverable stack status - Review the error, make changes in template and delete the stack to re-run the pipeline successfully; Reason: "
+          + getStackInfo(stack, "stackStatusReason");
+    }
```

Comment:
This is probably only related to stack operations (not change sets). Should this block be inside the `getStackInfo` method? In any case, this clearly calls for a refactor as you suggested to split the logic between stacks and changesets 👍

Reply:
I think this logic belongs to both cases, stack and changeset. 

*When the same scenario happened and a change set creation was requested, the error message was "NOT_YET_READY", which it does not reflect what actually happened.*

So when this happens, the flow goes through the change set logic and it returns NOT_YET_READY because there's not changeset information. 
So instead, we should return the stack status.

Reply:
Ah cool, makes sense. Thanks!